tests/e2e/scale/test_osd_node_balancing.py

```diff
@@ -0,0 +1,193 @@
+        new_osd_count = len(osd_running_nodes) + 3
+        pod_obj = ocp.OCP(kind=constants.POD, namespace="openshift-storage")
+        for new_osd_list in TimeoutSampler(
+            120, 30, lambda: pod_obj.get(selector=constants.OSD_APP_LABEL)["items"]
+        ):
+            if len(new_osd_list) == new_osd_count:
+                break
+    ceph_health_check(tries=30, delay=60)
+    collect_stats("OSD capacity has been increased")
+
+
+def collect_stats(action_text):
+    """
+    Write the current configuration information into the REPORT file.
+    This information includes the osd, nodes and which osds are on which
+    nodes.  The minimum and maximum numbers of osds per node are also
+    computed and saved.  If this is the final call to collect_stats
+    (action_text parameter is FINAL_REPORT), then the data collected
+    in the REPORT file is also displayed in the log.
+
+    Args:
+        action_text -- Title of last action taken
+                (usually adding nodes or adding osds)
+    """
+    pod_obj = ocp.OCP(
+        kind=constants.POD, namespace=constants.OPENSHIFT_STORAGE_NAMESPACE
+    )
+    osd_list = pod_obj.get(selector=constants.OSD_APP_LABEL)["items"]
+    node_stats = {}
+    for osd_ent in osd_list:
+        try:
+            osd_node = osd_ent["spec"]["nodeName"]
+        except KeyError:
+            continue
+        if osd_node in node_stats:
+            node_stats[osd_node].append(osd_ent)
+        else:
+            node_stats[osd_node] = [osd_ent]
+    osds_per_node = []
+    for entry in node_stats:
+        osds_per_node.append(len(node_stats[entry]))
+    wnodes = get_nodes(constants.WORKER_MACHINE)
+    for wnode in wnodes:
+        logging.info(wnode.name)
+        if wnode.name not in node_stats:
+            osds_per_node.append(0)
+    maxov = max(osds_per_node)
+    minov = min(osds_per_node)
+    this_skew = maxov - minov
+    logging.info(f"Skew found is {this_skew}")
+    logpath = ocsci_log_path()
+    outfile = f"{logpath}/{REPORT}"
+    with open(outfile, "a+") as fd:
+        fd.write(f"\nAction: {action_text}\n\n")
+        fd.write(f"Current OSDS: {len(osd_list)}\n")
+        for entry in osd_list:
+            fd.write(f"\t\t{entry['metadata']['name']}\n")
+        fd.write(f"\nCurrent Nodes: {len(wnodes)}\n")
+        for entry in wnodes:
+            fd.write(f"\t\t{entry.name}\n")
+        fd.write("\nOSD distribution:\n")
+        for entry in osd_list:
+            fd.write(f"\t\t{entry['metadata']['name']}")
+            try:
+                fd.write(f" is on {entry['spec']['nodeName']}\n")
+            except KeyError:
+                fd.write(" is on an unknown node\n")
+        fd.write(f"\nMaximum number of OSDs on a node: {maxov}\n")
+        fd.write(f"Minimum number of OSDs on a node: {minov}\n")
+        fd.write(f"Skew (Difference in OSD counts): {this_skew}\n")
+        if this_skew > 1:
+            fd.write("\n*** ERROR *** OSDs are not balanced between nodes\n")
```

Comment:
shouldn't we assert here?

Reply:
Yes.  I'll change this.

Reply:
Note: I think that this message should be changed to something other than ERROR.  I think that we should assert only when we add new nodes and when the test is finished.  The reason for this is when we have 3 nodes with 3 OSDs on each, when we add 3 new nodes we will have 3 nodes with 3 OSDs and 3 nodes with no OSDs.  Now when we evenly add an OSD to each of the nodes with 0 OSDs we will have 3 OSDs on 3 nodes and 1 OSD on 3 nodes.  This is the best that we can do in this case and we do not want to assert when this happens.

Reply:
Assert when imbalance occurs has been added.